packages/swapper/src/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts

```diff
@@ -225,7 +226,80 @@ async function _getZrxPermit2TradeQuote(
     )
   }
 
-  const maybeZrxPriceResponse = await fetchFromZrxPermit2({
+  // If we don't have a wallet, no dice for the permit2 EIP712 data here - but we don't care just yet since we're getting
+  // a *rate* quote, not a *quote* quote on quote quote.
+  if (!hasWallet) {
+    const maybeZrxPriceResponse = await fetchFromZrxPermit2({
+      priceOrQuote: 'price',
+      buyAsset,
+      sellAsset,
+      sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      receiveAddress,
+      affiliateBps,
+      slippageTolerancePercentageDecimal,
+      zrxBaseUrl,
+    })
+
+    if (maybeZrxPriceResponse.isErr()) return Err(maybeZrxPriceResponse.unwrapErr())
+    const zrxPriceResponse = maybeZrxPriceResponse.unwrap()
+
+    const {
+      buyAmount: buyAmountAfterFeesCryptoBaseUnit,
+      minBuyAmount: buyAmountBeforeFeesCryptoBaseUnit,
+      totalNetworkFee,
+    } = zrxPriceResponse
+
+    const rate = bnOrZero(buyAmountAfterFeesCryptoBaseUnit)
+      .div(sellAmountIncludingProtocolFeesCryptoBaseUnit)
+      .toString()
+
+    // 0x approvals are cheaper than trades, but we don't have dynamic quote data for them.
+    // Instead, we use a hardcoded gasLimit estimate in place of the estimatedGas in the 0x quote response.
+    try {
+      const networkFeeCryptoBaseUnit = totalNetworkFee
+      return Ok({
+        id: uuid(),
+        receiveAddress,
+        potentialAffiliateBps,
+        affiliateBps,
+        // Slippage protection is only provided for specific pairs.
+        // If slippage protection is not provided, assume a no slippage limit.
+        // If slippage protection is provided, return the limit instead of the estimated slippage.
+        // https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-quote
+        slippageTolerancePercentageDecimal,
+        rate,
+        steps: [
+          {
+            estimatedExecutionTimeMs: undefined,
+            // We don't care about this - this is a rate, and if we really wanted to, we know the permit2 allowance target
+            allowanceContract: undefined,
+            buyAsset,
+            sellAsset,
+            accountNumber,
+            rate,
+            feeData: {
+              protocolFees: {},
+              networkFeeCryptoBaseUnit, // L1 fee added inside of evm.calcNetworkFeeCryptoBaseUnit
+            },
+            buyAmountBeforeFeesCryptoBaseUnit,
+            buyAmountAfterFeesCryptoBaseUnit,
+            sellAmountIncludingProtocolFeesCryptoBaseUnit,
+            source: SwapperName.Zrx,
+          },
+        ] as unknown as SingleHopTradeQuoteSteps,
+      })
+    } catch (err) {
+      return Err(
+        makeSwapErrorRight({
+          message: 'failed to get fee data',
+          cause: err,
+          code: TradeQuoteError.NetworkFeeEstimationFailed,
+        }),
+      )
+    }
```

Comment:
preferably blocking: nothing to catch 😅

```suggestion
    const networkFeeCryptoBaseUnit = totalNetworkFee
    return Ok({
      id: uuid(),
      receiveAddress,
      potentialAffiliateBps,
      affiliateBps,
      // Slippage protection is only provided for specific pairs.
      // If slippage protection is not provided, assume a no slippage limit.
      // If slippage protection is provided, return the limit instead of the estimated slippage.
      // https://0x.org/docs/0x-swap-api/api-references/get-swap-v1-quote
      slippageTolerancePercentageDecimal,
      rate,
      steps: [
        {
          estimatedExecutionTimeMs: undefined,
          // We don't care about this - this is a rate, and if we really wanted to, we know the permit2 allowance target
          allowanceContract: undefined,
          buyAsset,
          sellAsset,
          accountNumber,
          rate,
          feeData: {
            protocolFees: {},
            networkFeeCryptoBaseUnit, // L1 fee added inside of evm.calcNetworkFeeCryptoBaseUnit
          },
          buyAmountBeforeFeesCryptoBaseUnit,
          buyAmountAfterFeesCryptoBaseUnit,
          sellAmountIncludingProtocolFeesCryptoBaseUnit,
          source: SwapperName.Zrx,
        },
      ] as unknown as SingleHopTradeQuoteSteps,
    })
```

Reply:
Will tackle in https://github.com/shapeshift/web/pull/7986 too which is an effective cleanup PR 🙏🏽

Reply:
gm 12b8327637